### PR TITLE
(prototype) One SimCLR example, one benchmark, and the gate between them

### PR DESCRIPTION
**Pull request by gabrielfruet**
> **Prototype. Do not merge.** This is a design exploration for the LightlySSL 2.0
> refactor, opened to be read and argued with, not to land on `master`. The whole
> stack goes together or not at all, and the shape is still open.

6 of 7 in a stack. Base: #2037.

SimCLR exists four times in this tree and nothing compares the copies. The two shipped heads differ by 4,922,112 parameters, so `load_state_dict` raises. The split and fused forwards give gradients at cosine similarity 0.0837. Both examples take `NTXentLoss()`'s default 0.5 while the benchmark passes 0.1.

`examples/simclr.py` is plain torch with the loop in view, small enough to run: ResNet-18, CIFAR-10, batch 256. `benchmarks/simclr/` is the same method on Lightning, parameterised by one row of `datasets.py`, where the `imagenet` row is the paper's numbers. Every row states every field, so two rows side by side show every difference.

`tests/test_simclr_agrees.py` is what makes them one method. It compares the method, not the run: block classes, the view contract, the fused forward, the head's width against the backbone's, and that each file's criterion carries the temperature that file declares. Batch size, backbone, learning rate and temperature values are free to differ.

The old four are deleted in PR 7, not here.

Testing: 14 gate cases; mutating either side's forward to run one view at a time fails the gate and names the side.

### CODE REVIEW

File: examples/simclr.py [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+criterion = NTXentLoss(temperature=TEMPERATURE)
+transform = SimCLRTransform(
+    input_size=INPUT_SIZE,
+    cj_strength=COLOR_JITTER_STRENGTH,
+    gaussian_blur=GAUSSIAN_BLUR,
+    normalize=NORMALIZE,
+)
+
+
+def forward(sample: Sample) -> Tensor:
```
Comment (gabrielfruet) [inline examples/simclr.py]: Use nn.Module instead here. For PyTorch users this will look very weird

File: examples/simclr.py [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+from lightly.models.modules import SimCLRProjectionHead
+from lightly.optim import param_groups
+from lightly.transforms import SimCLRTransform
+from lightly.transforms.utils import CIFAR10_NORMALIZE
+
+# ResNet-18, CIFAR-10, 32 pixels. Small enough to run on one GPU.
+BATCH_SIZE, EPOCHS, LR, MOMENTUM = 256, 100, 0.06, 0.9
+TEMPERATURE, WEIGHT_DECAY = 0.5, 5e-4
+INPUT_SIZE, FEATURE_DIM, HIDDEN_DIM, OUT_DIM = 32, 512, 512, 128
+COLOR_JITTER_STRENGTH, GAUSSIAN_BLUR, NORMALIZE = 0.5, 0.0, CIFAR10_NORMALIZE
```
Comment (gabrielfruet) [inline examples/simclr.py]: The example should be clearer. The thing is that the user should be able to **run it**, not exactly match the paper parameter.

For the benchmarks, it should be the "same as the paper"

File: benchmarks/simclr/benchmark.py [hunk 3]
```diff
@@ -0,0 +1,248 @@
...
+    """Builds the row's transform. Restated from the example; the gate compares it."""
+    return SimCLRTransform(
+        input_size=setting.size,
+        cj_strength=setting.color_jitter_strength,
+        gaussian_blur=setting.blur,
+        normalize=setting.normalize,
+    )
+
+
+class SimCLR(LightningModule):
```
Comment (gabrielfruet) [inline benchmarks/simclr/benchmark.py]: Benchmarks are a bit hidden now. They should have the source of truth for the parameters

"Anyone that is serious about running the benchmarks could easily look into github"

We need to think of a better way of saving the parameters (maybe exposing in the docs in someway, or making it easily accesible.

File: benchmarks/simclr/benchmark.py [hunk 4]
```diff
@@ -0,0 +1,248 @@
...
+def transform(setting: Setting) -> SimCLRTransform:
+    """Builds the row's transform. Restated from the example; the gate compares it."""
+    return SimCLRTransform(
+        input_size=setting.size,
+        cj_strength=setting.color_jitter_strength,
+        gaussian_blur=setting.blur,
+        normalize=setting.normalize,
+    )
+
+
```
Comment (gabrielfruet) [inline benchmarks/simclr/benchmark.py]: Maybe call the Examples -> Examples and Benchmarks and show the PyTorch and the Benchmark.
